File: tests/polymer/TorsionAngle.cpp
#include <gtest/gtest.h>

#include "xmol/polymer/TorsionAngle.h"
#include "xmol/polymer/exceptions.h"
#include "xmol/pdb/PdbFile.h"
#include "xmol/pdb/PdbReader.h"
#include <cmath>

using ::testing::Test;
using namespace xmol::geometry;
using namespace xmol::polymer;
using namespace xmol::pdb;

class TorsionAngleTests : public Test {
public:
};

TEST_F(TorsionAngleTests, test1) {
  Frame frame(1);
  Chain& c = frame.emplace(ChainName("A"));
  Residue& r = c.emplace(ResidueName("LYS"), residueId_t(1));
  r.emplace(AtomName("A"), atomId_t(0), XYZ(1, 0, 0));
  r.emplace(AtomName("B"), atomId_t(1), XYZ(0, 0, 0));
  r.emplace(AtomName("C"), atomId_t(2), XYZ(0, 0, 1));
  r.emplace(AtomName("D"), atomId_t(3), XYZ(0, 1, 1));

  auto t = TorsionAngle(r[AtomName("A")], r[AtomName("B")], r[AtomName("C")], r[AtomName("D")],
                        [](Atom& a, Atom& b, Atom& c, Atom& d) {
                          Atom* atoms[] = {&d};
                          return AtomSelection(std::begin(atoms), std::end(atoms));
                        });
  EXPECT_DOUBLE_EQ(t.value().degrees(), 90);

  t.set(Degrees(90)); // noop
  EXPECT_DOUBLE_EQ(t.value().degrees(), 90);

  t.set(Degrees(0));
  EXPECT_TRUE(fabs(t.value().degrees() - 0) < 1e-10);

  EXPECT_TRUE(fabs(r[AtomName("D")].r().x() - 1) < 1e-10);
  EXPECT_TRUE(fabs(r[AtomName("D")].r().y() - 0) < 1e-10);
  EXPECT_TRUE(fabs(r[AtomName("D")].r().z() - 1) < 1e-10);

  auto throws1 = [&] {
    auto t2 = TorsionAngle(r[AtomName("A")], r[AtomName("B")], r[AtomName("C")], r[AtomName("D")],
                           [](Atom& a, Atom& b, Atom& c, Atom& d) {
                             Atom* atoms[] = {&a, &d};
                             return AtomSelection(std::begin(atoms), std::end(atoms));
                           });
  };

  EXPECT_THROW(throws1(), PolymerException);
  auto throws2 = [&] {
    auto t3 = TorsionAngle(r[AtomName("A")], r[AtomName("B")], r[AtomName("C")], r[AtomName("D")],
                           [](Atom& a, Atom& b, Atom& c, Atom& d) {
                             Atom* atoms[] = {&b, &c};
                             return AtomSelection(std::begin(atoms), std::end(atoms));
                           });
  };

  EXPECT_THROW(throws2(), PolymerException);

  auto t4 = TorsionAngle(r[AtomName("A")], r[AtomName("B")], r[AtomName("C")], r[AtomName("D")]);
  EXPECT_TRUE(fabs(t4.value().degrees() - 0) < 1e-10);

  EXPECT_THROW(t4.set(Degrees(10)), PolymerException);


}
TEST_F(TorsionAngleTests, test_backbone_angles){
  auto files ={
      "pdb/rcsb/1PGB.pdb",
      "pdb/rcsb/5BMG.pdb",
      "pdb/rcsb/5BMH.pdb",
  };
  for(auto& file: files){
    auto frame = PdbFile(file).get_frame();

    for(auto&r : frame.asResidues()){

      auto phi = TorsionAngleFactory::phi(r);
      auto psi = TorsionAngleFactory::psi(r);
      auto omega = TorsionAngleFactory::omega(r);
      if (!!phi){
        phi.value().set(Degrees(180.0));
      }
      if (!!psi){
        psi.value().set(Degrees(180.0));
      }
      if (!!omega){
        omega.value().set(Degrees(180.0));
      }
    }
  }
}

File: docs/source/examples/basics/torsion_angle.py
"""

.. |Atom| replace:: :py:class:`~pyxmolpp2.polymer.Atom`
.. |Residue| replace:: :py:class:`~pyxmolpp2.polymer.Residue`
.. |Chain| replace:: :py:class:`~pyxmolpp2.polymer.Chain`
.. |Frame| replace:: :py:class:`~pyxmolpp2.polymer.Frame`

.. |AtomSelection| replace:: :py:class:`~pyxmolpp2.polymer.AtomSelection`
.. |ResidueSelection| replace:: :py:class:`~pyxmolpp2.polymer.ResidueSelection`
.. |ChainSelection| replace:: :py:class:`~pyxmolpp2.polymer.ChainSelection`

.. |AtomPredicate| replace:: :py:class:`~pyxmolpp2.polymer.AtomPredicate`
.. |ResiduePredicate| replace:: :py:class:`~pyxmolpp2.polymer.ResiduePredicate`
.. |ChainPredicate| replace:: :py:class:`~pyxmolpp2.polymer.ChainPredicate`



Torsion Angles
--------------

"""

import pyxmolpp2
import os

##############################################################################
# Let's create a frame to work with

pdb_filename = os.path.join(os.environ["TEST_DATA_PATH"], "pdb/rcsb/1UBQ.pdb")
pdb_file = pyxmolpp2.pdb.PdbFile(pdb_filename)

frame = pdb_file.get_frame()

##############################################################################
# Standard torsion angles
# ^^^^^^^^^^^^^^^^^^^^^^^
#
# For standard protein residues angles can be constructed using `TorsionAngleFactory`:

from pyxmolpp2.polymer import TorsionAngleFactory
from pyxmolpp2.geometry import Degrees

residue48 = frame.asChains[0][48]
print(residue48)

psi_48 = TorsionAngleFactory.psi(residue48)
print(psi_48)
print(psi_48.value().degrees)

##############################################################################
# Note: Factory may return ``None`` if such angle does not exist:
print(TorsionAngleFactory.omega(frame.asResidues[0]))

##############################################################################
# Torsion angle allows to set a new one:

# All residues 49-76 are affected by this rotation
psi_48.set(Degrees(150))
print(psi_48.value().degrees)

##############################################################################
# Construction
# ^^^^^^^^^^^^
# Torsion angle constructor allow two forms:
#  1. Read-only torsion angle
#  2. Read-write torsion angle

from pyxmolpp2.polymer import TorsionAngle, AtomName, Atom

r1 = frame.asResidues[1]
r2 = frame.asResidues[2]

# Let's create a read-only phi of residue 2
phi_2_ro = TorsionAngle(r1[AtomName("C")],
                        r2[AtomName("N")],
                        r2[AtomName("CA")],
                        r2[AtomName("C")],
                        )

# Check against factory angle:
assert phi_2_ro.value().degrees == TorsionAngleFactory.phi(r2).value().degrees

##############################################################################
# Attempt to set TorsionAngle will lead to ``RuntimeError``:
try:
    phi_2_ro.set(Degrees(-130))
except RuntimeError as e:
    print(e)


##############################################################################
# We need a helper function which returns a selection of affected atoms
# by our torsion angle

def affected_phi_atoms(a: Atom, b: Atom, c: Atom, d: Atom):
    from pyxmolpp2.polymer import rId
    return a.chain.asResidues.filter(rId > a.rId).asAtoms


phi_2_rw = TorsionAngle(r1[AtomName("C")],
                        r2[AtomName("N")],
                        r2[AtomName("CA")],
                        r2[AtomName("C")],
                        affected_phi_atoms
                        )

phi_2_rw.set(Degrees(-130))

print(phi_2_ro.value().degrees)
print(phi_2_rw.value().degrees)


